src/pykeen/models/predict.py

```diff
@@ -0,0 +1,465 @@
+        ],
+        columns=['tail_id', 'tail_label', 'score'],
+    ).sort_values('score', ascending=False)
+
+    return _postprocess_prediction_df(
+        rv,
+        add_novelties=add_novelties,
+        remove_known=remove_known,
+        testing=testing,
+        training=model.triples_factory.mapped_triples,
+        query_ids_key='tail_id',
+        col=2,
+        other_col_ids=(head_id, relation_id),
+    )
+
+
+def predict_scores_all_relations(
+    model: Model,
+    ht_batch: torch.LongTensor,
+    slice_size: Optional[int] = None,
+) -> torch.FloatTensor:
+    """Forward pass using middle (relation) prediction for obtaining scores of all possible relations.
+
+    This method calculates the score for all possible relations for each (head, tail) pair.
+
+    Additionally, the model is set to evaluation mode.
+
+    :param ht_batch: shape: (batch_size, 2), dtype: long
+        The indices of (head, tail) pairs.
+    :param slice_size: >0
+        The divisor for the scoring function when using slicing.
+
+    :return: shape: (batch_size, num_relations), dtype: float
+        For each h-t pair, the scores for all possible relations.
+    """
+    # Enforce evaluation mode
+    model.eval()
+    if slice_size is None:
+        scores = model.score_r(ht_batch)
+    else:
+        scores = model.score_r(ht_batch, slice_size=slice_size)
+    if model.predict_with_sigmoid:
+        scores = torch.sigmoid(scores)
+    return scores
+
+
+def score_all_triples(
+    model: Model,
+    k: Optional[int] = None,
+    batch_size: int = 1,
+    return_tensors: bool = False,
+    add_novelties: bool = True,
+    remove_known: bool = False,
+    testing: Optional[torch.LongTensor] = None,
+) -> Union[Tuple[torch.LongTensor, torch.FloatTensor], pd.DataFrame]:
+    """Compute scores for all triples, optionally returning only the k highest scoring.
+
+    .. note:: This operation is computationally very expensive for reasonably-sized knowledge graphs.
+    .. warning:: Setting k=None may lead to huge memory requirements.
+
+    :param k:
+        The number of triples to return. Set to None, to keep all.
+
+    :param batch_size:
+        The batch size to use for calculating scores.
+
+    :return: shape: (k, 3)
+        A tensor containing the k highest scoring triples, or all possible triples if k=None.
+
+    Example usage:
+
+    .. code-block:: python
+
+        from pykeen.pipeline import pipeline
+
+        # Train a model (quickly)
+        result = pipeline(model='RotatE', dataset='Nations', training_kwargs=dict(num_epochs=5))
+        model = result.model
+
+        # Get scores for *all* triples
+        tensor = model.score_all_triples()
+        df = model.make_labeled_df(tensor)
+
+        # Get scores for top 15 triples
+        top_df = model.score_all_triples(k=15)
+    """
+    # set model to evaluation mode
+    model.eval()
+
+    # Do not track gradients
+    with torch.no_grad():
```

Comment:
`torch.no_grad()` can also be used as function annotation, and helps to reduce the indention here

Reply:
Nice hint, thanks! I did that after refactoring

Reply:
I wonder why this is necessary when you're in model.eval() mode though